AGENT-51: Add documentation for preset configurations (#1388)
openshift-metal3/dev-scripts: agent/docs/config-presets-and-options.md

# Configuration Presets and Options

> ## Defining the Openshift Cluster Topology

Reviewer: ```suggestion
## Defining the OpenShift Cluster Topology
```

> Openshift clusters can have different topologies. In Openshift development we refer to 

Reviewer: ```suggestion
OpenShift clusters can have different topologies. In OpenShift development we refer to 
```

these different topologies as different installations with adjectives and acronyms. 

> Below is the list of cluster toplogies currently being used to test the installers:

Reviewer: ```suggestion
Below is the list of cluster topologies currently being used to test the installers:
```

| NUM_WORKERS         |    2    |           2+          |    0    |              0              |
| NUM_EXTRA_WORKERS   |    0    |           0           |    0    |              0              |
| MASTER_MEMORY       |  16384  |         16384         |  16384  |            32768            |
| MASTER_DISK         |    30   |          120          |   120   |             120             |
| MASTER_VCPU         |    8    |           8           |    8    |              8              |
| WORKER_MEMORY       |   8192  |         16384         |         |                             |
| WORKER_DISK         |    30   |          120          |         |                             |
| WORKER_VCPU         |    4    |           4           |         |                             |
| EXTRA_WORKER_MEMORY |   8192  |         16384         |         |                             |
| EXTRA_WORKER_DISK   |    30   |          120          |         |                             |
| EXTRA_WORKER_VCPU   |    4    |           4           |         |                             |

## Defining the Network Topology

> The different network topologies we currently use while testing Openshift installations are defined below.

Reviewer: ```suggestion
The different network topologies we currently use while testing OpenShift installations are defined below.
```

| MIRROR_IMAGES[1]       |            true           |                  |            true           |            true           |
| HOST_IP_STACK[2]       |             v6            |        v4        |             v6            |            v4v6           |
| PROVISIONING_NETWORK   |      fd00:1101::0/64      |   172.22.0.0/24  |      fd00:1101::0/64      |      fd00:1101::0/64      |
| EXTERNAL_SUBNET_V4     |                           | 192.168.111.0/24 |                           |      192.168.111.0/24     |
| EXTERNAL_SUBNET_V6     | fd2e:6f44:5dd8:c956::/120 |                  | fd2e:6f44:5dd8:c956::/120 | fd2e:6f44:5dd8:c956::/120 |

* [1]: MIRROR_IMAGES is unset unless except in the case of IPv6. The true default value of MIRROR_IMAGES is that it is unset.
* [2]: If HOST_IP_STACK is not set then the value is equal to the value of IP_STACK.

### Static IP vs DHCP

TODO(lranjbar): Currently this seems to be only set in our Agent flows for generating static IP. (?)

### Connected vs Disconnected

By default all clusters created by dev-scripts are connected clusters. Meaning the cluster is 

> connected to the greeater internet. In order to simular a disconnected cluster installation we first

Reviewer: ```suggestion
connected to the greater internet. In order to simulate a disconnected cluster installation we first
```

create a local mirrored image registry the same way a customer would in disconnected installation.

> Currently Quay.io does not support IPv6. So these IPv6 installations are disconnected since we have to

Reviewer: ```suggestion
Currently quay.io does not support IPv6. So these IPv6 installations are disconnected since we have to
```

# Configuration Presets and Options

## Defining the Openshift Cluster Topology

Openshift clusters can have different topologies. In Openshift development we refer to 
these different topologies as different installations with adjectives and acronyms. 
Below is the list of cluster toplogies currently being used to test the installers:

|                       | Master Nodes | Worker Nodes | Masters Schedulable |
|-----------------------|--------------|--------------|---------------------|
| High Availablity (HA) |       3+     |       2+     |         No          |
| Compact               |       3      |       0      |         Yes         |
| Single Node Openshift |       1      |       0      |         Yes         |

### Minimum Cluster Resources

|                     | High Availablity (HA) | Compact | Single Node OpenShift (SNO) |
|---------------------|-----------------------|---------|-----------------------------|
| NUM_MASTERS         |           3+          |    3    |              1              |

> | NUM_WORKERS         |           2+          |    0    |              0              |

Author: The table above is more on an explanation table... this one is describing the actual environment variables in dev-scripts. I'll a line explaining that?

### IP Stack

|                        |       IPv4       |            IPv6           |       IP Dual-stack       |
|------------------------|------------------|---------------------------|---------------------------|
| IP_STACK               |        v4        |             v6            |            v4v6           |
| NETWORK_TYPE           |   OpenShiftSDN   |       OVNKubernetes       |       OVNKubernetes       |
| CLUSTER_SUBNET_V4      |   10.128.0.0/14  |                           |       10.128.0.0/14       |
| CLUSTER_SUBNET_V6      |                  |         fd01::/48         |         fd01::/48         |
| CLUSTER_HOST_PREFIX_V4 |        23        |                           |             23            |
| CLUSTER_HOST_PREFIX_V6 |                  |                           |             64            |
| SERVICE_SUBNET_V4      |                  |                           |       172.30.0.0/16       |
| SERVICE_SUBNET_V6      |                  |         fd02::/112        |         fd02::/112        |
| MIRROR_IMAGES          |                  |            true           |            true           |

> | HOST_IP_STACK          |        v4        |             v6            |            v4v6           |

Author: If HOST_IP_STACK is not set it defaults to whatever IP_STACK is set to.

> | HOST_IP_STACK          |        v4        |             v6            |            v4v6           |

Author: ...and the default for IP_STACK is v4.

> | HOST_IP_STACK          |        v4        |             v6            |            v4v6           |

Author: Should I add a column for default values maybe?
